tket/src/Circuit/CircUtils.cpp

```diff
@@ -717,6 +710,331 @@ Circuit with_controls(const Circuit &c, unsigned n_controls) {
   return c2;
 }
 
+// Return the target unitary given a Cn* gate where n >= 0
+static Eigen::Matrix2cd get_target_op_matrix(const Op_ptr &op) {
+  OpType optype = op->get_type();
+  Eigen::Matrix2cd m;
+  switch (optype) {
+    case OpType::CX:
+    case OpType::CCX:
+    case OpType::CnX:
+      return Gate(OpType::X, {}, 1).get_unitary();
+    case OpType::CSX:
+      return Gate(OpType::SX, {}, 1).get_unitary();
+    case OpType::CSXdg:
+      return Gate(OpType::SXdg, {}, 1).get_unitary();
+    case OpType::CV:
+      return Gate(OpType::V, {}, 1).get_unitary();
+    case OpType::CVdg:
+      return Gate(OpType::Vdg, {}, 1).get_unitary();
+    case OpType::CRx:
+      return Gate(OpType::Rx, op->get_params(), 1).get_unitary();
+    case OpType::CnRy:
+    case OpType::CRy:
+      return Gate(OpType::Ry, op->get_params(), 1).get_unitary();
+    case OpType::CY:
+      return Gate(OpType::Y, {}, 1).get_unitary();
+    case OpType::CRz:
+      return Gate(OpType::Rz, op->get_params(), 1).get_unitary();
+    case OpType::CZ:
+      return Gate(OpType::Z, {}, 1).get_unitary();
+    case OpType::CH:
+      return Gate(OpType::H, {}, 1).get_unitary();
+    case OpType::CU1:
+      return Gate(OpType::U1, op->get_params(), 1).get_unitary();
+    case OpType::CU3:
+      return Gate(OpType::U3, op->get_params(), 1).get_unitary();
+    default:
+      if (!is_gate_type(optype) || op->n_qubits() != 1) {
+        throw CircuitInvalidity(
+            "Cannot get the target unitary of " + op->get_name());
+      }
+      return as_gate_ptr(op)->get_unitary();
+  }
+}
+
+// A gate block containing Cn* gates that can be merged as a single CnU gate
+struct CnGateBlock {
+  enum class MergeMode { append, prepend };
+  CnGateBlock(const Command &command) {
+    // Assumes the color of the target is not identity
+    Op_ptr op = command.get_op_ptr();
+    ops.push_back(op);
+    unit_vector_t args = command.get_args();
+    TKET_ASSERT(!args.empty());
+    for (unsigned i = 0; i < args.size() - 1; i++) {
+      control_qubits.insert(args[i].index()[0]);
+    }
+    target_qubit = args.back().index()[0];
+    is_symmetric =
+        (op->get_type() == OpType::CZ || op->get_type() == OpType::CU1);
+    color = as_gate_ptr(op)->commuting_basis(args.size() - 1);
+    TKET_ASSERT(color != Pauli::I);
```

Comment:
I'm not sure we can actually assert this, because `commuting_basis()` can return `Pauli::I`. Should we throw an exception in this case, or should we handle it?

Reply:
As far as I know only ``BRIDGE`` and ``noop`` have ``Pauli::I`` ports. Since ``CnGateBlock`` is only being used in ``with_controls_numerical``,  bridges should have been rebased into ``CXs`` and ``noop `` should have been ignored. But I think an exception is more suitable here.